File: Dsp-lab/Experiment-4/codes/main.m
function main()

x = [0.3426 3.5784 2.7694 -1.3499 3.0349 0.7254 -0.0631];
h = [0.7147 -0.2050 -0.1241 1.4897 1.4090];

disp('convolution x and h by fix-point arithmetic is given by');
disp(fix_convl(x,h));

disp('corss-correlation of x and h by fix-point arithmetic is given by');
disp(fix_xcorrl(x,h));


disp('error of convolution from my function');
err_convl = mserr(fix_convl(x,h),float_convl(x,h),length(x)+length(h)-1);
disp(err_convl)

disp('error of correlation from my function');
err_xcorrl = mserr(fix_xcorrl(x,h),float_xcorrl(x,h),length(x)+length(h)-1);
disp(err_xcorrl);

%Plot
x = 1:1:length(x)+length(h)-1;
figure;
plot(x,err_convl,"g")
grid on;
figure;
plot(x,err_xcorrl,"b")
grid on;
end

function y = float_convl(x,h) % y here is the output

  l = length(x) + length(h) - 1;

  % Initialize output sequence
  y = zeros(1, l);

  % Perform convolution manually
  for n = 1:l
    for k = 1:length(x)
      if (n - k + 1) >= 1 && (n - k + 1) <= length(h) 
        y(n) = y(n) + x(k) * h(n - k + 1); % This is just /sigma x(k)*h(n-k)
      end
    end
  end
end

function y = fix_convl(x,h) % y here is the output
  
  % Calculate output length
  l = length(x) + length(h) - 1;

  % Initialize output sequence
  y = zeros(1, l);
  Q = 12;
  % Perform convolution manually
  for n = 1:l
    for k = 1:length(x)
      if (n - k + 1) >= 1 && (n - k + 1) <= length(h)
        
        yf = fixpoint(y(n),Q); 
        
        xf = fixpoint(x(k),Q); %x(k) taken in fixpoint
        
        hf = fixpoint(h(n - k + 1),Q); %x(n-k) taken in fixpoint
        
        xhf = xf*hf; %x(k)*h(n-k) product taken in fixpoint

        xh = xhf/(2^(2*Q)); %x(k)*h(n-k) product converted back  

        xhff = fixpoint(xh,Q); %x(k)*h(n-k)=xh taken in fixpoint   

        yf = yf + (xhff); % This is just /sigma x(k)*h(n-k) done in fixpoint

        y(n) = yf/(2^(Q)); % convolution in fixpoint converted back 
      end
    end
  end
end

function x0 = flip_lr(x)
    l = length(x);
    x0 = zeros(1,l);
    for i = 1:l
        x0(l-i+1) = x(i);
    end
end 

function y1 = fix_xcorrl(x,h) % y1 here is the output
  
  h0=flip_lr(h); %to rearange h(n) as h(-n)
  y1 = fix_convl(x,h0);

end

% cross_correlation

function y1 = float_xcorrl(x,h) % y1 here is the output
  
  h0=fliplr(h); %to rearange h(n) as h(-n)
  y1 = float_convl(x,h0);

end

function result = fixpoint(a, Q)
    result = fix(a * 2^Q);
end

function err = mserr(y_obs,y_acc,n)

err = zeros(1,n);

    for i = 1:n

        err(i) = (y_obs(i) - y_acc(i))^2 ;

    end
    disp('mse is')
    disp(sum((err))/(n));
    

end

File: Dsp-lab/Experiment-4/codes/main.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int fixpoint(float a, int Q)
{
    int y = round(a * pow(2, Q));
    return y;
}

float float_calc_conv(float *x, float *h, int l1, int l2,int index) // function to compute convolution
{
    int l = l1 + l2 - 1;
    float y[l];
    for (int n = 0; n < l; n++)
    {
        y[n] = 0;
        for (int k = 0; k < l1; k++)
        {
            if (n - k >= 0 && n - k < l2) 
            {
                y[n] = y[n] + x[k] * h[n  - k]; 
            }
        }
    }

    return y[index];
}

float fix_calc_conv(float *x, float *h, int l1, int l2,int index) // function to compute convolution
{
    int l = l1 + l2 - 1;
    
    float y[l],xh;
    
    int yf,xf,hf,xhf,xhff;

    for (int n = 0; n < l; n++)
    {
        y[n] = 0;
        int Q = 12;
        for (int k = 0; k < l1; k++)
        {
            if (n - k >= 0 && n - k < l2) /*keeping the h(n-k) sequnce same even though it has been folded and shifted */
            {
                yf = fixpoint(y[n],Q);
                xf = fixpoint(x[k],Q);
                hf = fixpoint(h[n-k],Q);

                xhf = xf*hf;
                xh = xhf/(pow(2,2*Q));

                xhff = fixpoint(xh,Q); 

                yf = yf + xhff;  
                y[n] = yf/(pow(2,Q));
            }
        }
    }

    return y[index];
}

float fliplr(float *x, int l1, int index) // function for folding the sequence i.e finding x(-n)
{
    float x0[l1];

    for (int i = 0; i < l1; i++)
    {
        x0[(l1 - 1) - i] = x[i];
    }
    return x0[index];
}

float fix_calc_corr(float *x, float *h, int l1, int l2,int index) // function to compute corss correlation
{
    float h0[l2];
    int l = l1 + l2 - 1;
    float y[l];
    for (int i = 0; i < l2; i++)
    {
        h0[i] = fliplr(h, l2, i);
    }
    for (int i = 0; i < l1 + l2 - 1; i++)
    {
        y[i]=fix_calc_conv(x, h0, l1, l2,i);
    }
    return y[index];
    
}

float float_calc_corr(float *x, float *h, int l1, int l2,int index) // function to compute corss correlation
{
    float h0[l2];
    int l = l1 + l2 - 1;
    float y[l];
    for (int i = 0; i < l2; i++)
    {
        h0[i] = fliplr(h, l2, i);
    }
    for (int i = 0; i < l1 + l2 - 1; i++)
    {
        y[i]=float_calc_conv(x, h0, l1, l2,i);
    }
    return y[index];
    
}

float error(float *x_obs,float *x_acc,int n,int index) 
{
    float err[n];

    for(int i=0;i<n;i++)
    {
        err[i]=(pow((x_obs[i]-x_acc[i]),2));
        
    }
    return err[index];
}

float sum(float *x, int n) {
    
    float total = 0;

    // Loop through the array using a pointer, efficiently accessing elements
    for (int i = 0; i < n; i++) {
        total += *(x + i); // Dereference the pointer with offset to get each element
    }

    float avg = total/n;

    return avg;
}

int main() 
{
    int l1 = 7;
    int l2 = 5;
    int l =l1 +l2 -1;
    float x[7] = {0.3426, 3.5784, 2.7694, -1.3499, 3.0349, 0.7254, -0.0631};
    float h[5] = {0.7147, -0.2050, -0.1241, 1.4897, 1.4090};
    
    printf("Convolution by fixpoint arithemetic of x and h resulted:\n");
    
    for (int i = 0; i < l1 + l2 - 1; i++)
    {
        printf("%.4f   ", fix_calc_conv(x, h, l1, l2,i));
    }
    printf("\n");

    printf("Convolution by floatpoint arithemetic of x and h resulted:\n");

    for (int i = 0; i < l1 + l2 - 1; i++)
    {
        printf("%.4f   ", float_calc_conv(x, h, l1, l2,i));
    }
    printf("\n");

    printf("Cross-correlation by fixpoint arithemetic of x and h resulted:\n");
    
    for (int i = 0; i < l1 + l2 - 1; i++)
    {
        printf("%.4f   ", fix_calc_corr(x, h, l1, l2,i));
    }
    printf("\n");

    printf("Cross-correlation floatpoint arithemetic of x and h resulted:\n");

    for (int i = 0; i < l1 + l2 - 1; i++)
    {
        printf("%.4f   ", float_calc_corr(x, h, l1, l2,i));
    }
    printf("\n");

    printf("the mean square error for convolution \n");

    // for computing mse in convolution
    float y_acc[l],y_obs[l],err[l];

    for(int i = 0; i < l1 + l2 - 1; i++)
    {
        y_obs[i]=float_calc_conv(x, h, l1, l2,i);
        y_acc[i]=fix_calc_conv(x, h, l1, l2,i);
    }

    for (int i = 0; i < l1 + l2 - 1; i++)
    {
        err[i]=error(y_obs,y_acc,l,i);
        printf("%E   ", error(y_obs,y_acc,l,i));
    }
    printf("\n");

    printf("The mse sum is %E",sum(err,l));

    printf("\n");

    // for computing mse in correlation

    printf("the mean square error for correlation\n");

    for(int i = 0; i < l1 + l2 - 1; i++)
    {
        y_obs[i]=float_calc_corr(x, h, l1, l2,i);
        y_acc[i]=fix_calc_corr(x, h, l1, l2,i);
    }    

    for (int i = 0; i < l1 + l2 - 1; i++)
    {
        err[i]=error(y_obs,y_acc,l,i);
        printf("%E   ", error(y_obs,y_acc,l,i));
    }
    printf("\n");

    printf("The mse sum is %E",sum(err,l));
    
    printf("\n");

    return 0;
}